### ISSUE TITLE
installing and dlib

### ISSUE THREAD

**Issue by bendudson** [nibbula/lish#1]
Hi thanks for making this! I was hoping something like this existed, or was thinking about writing one. Is there anything I can help with? 

I'm trying to install, and am having trouble finding the dlib dependency: It's not on Quicklisp, and Google unhelpfully suggests a machine learning library. Where could I find it?

**Comment by nibbula** [comment]
Oh my! Thank you for your interest in Lish. I suppose this had to happen
eventually. Unfortunately Lish is *totally* not ready for release, or even
probably use for anyone besides me. Not only are the dependencies
unnecessarily a mess, but I have planned to rewrite, very likely incompatibly,
every major piece of it, now that I have good idea of the issues involved.

That being said, I have been using it as my primary shell for years, and
it does have about 90% of the features I want from it, barring bugs. Unfortunately, the
remaining 5-10% is always the hardest and really the most crucial in taking it
from a toy to a thing that can truly be widely useful.

If you are interested in trying to cobble together a version for testing or
playing with, either I could send you a pre-baked binary, or I could try to
put together short instructions on how you could load/run/compile it from source.
Is there a platform you would prefer? I have had it working at times on Linux,
 macOS, FreeBSD, Solaris, etc. and on most CL implementations. I'm even about half done with Windows porting.

As far as helping, you would be the first person to try it, so just your
experience of running it, and any feedback or criticism, would surely be
quite helpful. Of course you're also quite welcome to work on or fix any part
of it. I'm just now getting back from a short hiatus from working on it, and
am currently trying to figure out what my next steps will be.

One major caveat! Please don't use it like a shell around any data you care to
keep. It's about as tested as an uneaten sandwich. I am quite aware it is
rife with bugs, perhaps of the file clobbering kind. I have heard Brian Fox,
the primary author of bash, describe how he deleted much of an early version
of his code while testing the shell in development, due to bugs in the globbing
code. I happen to know that I also have a few bugs in the globbing code.
So, please be cautious.

TL;DR The dependencies are in:
https://github.com/nibbula/useless-pile-of-junk-with-a-catchy-name
https://github.com/nibbula/opsys
https://github.com/nibbula/rl

I theory you could just toss these into your ~/quicklisp/local-projects (or
somewhere in your ASDF path) and give it a try.

Some other external dependencies, which are thankfully in quicklisp, are likely:
    - cl-ppcre
    - trivial-gray-streams
    - cffi

**Comment by bendudson** [comment]
Thank you for this, it's very helpful. It works! (for some value of works) If possible I would much rather help improve something than write another competing and probably inferior version. If you are thinking about refactoring then I'll post an issue before starting any major modifications.

I'm running Debian Jessie with SBCL 1.4.2. I cloned your `rl`, `opsys`, and `useless-pile-of-junk-with-a-catchy-name` (great name) repos into quicklisp `local-projects` and it (mostly) loads. Thanks for the warning; I'm using a throwaway user account for this initial testing. 

My recipe so far is:

0. Don't use SLIME, but run from SBCL directly. If running from SLIME then `:ansi` and `:ansi-stream` terminal types enter infinite loops, with either `posix-open: No such device or address` or `The slot TERMINAL::OUTPUT-STREAM is unbound in the object` errors printed repeatedly.

1. Comment out `find-slot-name`, `+simple-condition-format-control-slot+` and `+simple-condition-format-arguments-slot+` in `dlib1.lisp`, otherwise I get an error
``` 
Package MOP does not exist.
;   
;       Line: 1467
```

2. Load lish `(ql:quickload :lish)`

3. Uncomment lines in `dlib1.lisp` and evaluate or reload

4. Set the terminal type: `(setq lish::*default-terminal-type* :ansi)`
or I get an error `There is no class named COMMON-LISP:NIL.` coming from `(SB-PCL::FIND-CLASS-FROM-CELL NIL NIL T)` via `rl:initialize-instance` in `editor.lisp` (method starts line 229) when evaluating `(lish:lish)`. 

5. Run! `(lish:lish)` and get 
```
Couldn't find RC file: "/home/ben/.lishrc"
Lish>
```
Basic commands (ls, cd etc) work, tab completion is great. Evaluating some simple expressions and defining functions seems to work fine. I particularly like the line continuation if an expression hasn't closed all parentheses. Cool! Thanks very much.

The MOP package issue is an odd error, as I can see it should be defined in that same file. Unfortunately I don't know enough about MOP (or anything really). Is the approach you use similar to `closer-mop`?

**Comment by bendudson** [comment]
Thanks, those fixes work on my setup, and building an executable works. One small thing: the libmagic binding currently needs `libmagic-dev` installed under Debian, since `libmagic1` installs a file `libmagic.so.1` rather than the `libmagic.so` which Lish expects.

**Comment by nibbula** [comment]
I made a patch that should fix the libmagic issue.
Lish should also now be able to at least sort of work in "M-x shell" in emacs, if you ignore the redundant echoing of the command line. It's still a little bit away from it working in SLIME, and unless I do a rather detailed "mash-up" of it's features with SLIME, it probably won't work well. Cheers!
